### PR TITLE
PIMOB-1906: Enable optional CVV

### PR DESCRIPTION
**Pull request by alex-ioja-yang-cko**
[Jira task](https://checkout.atlassian.net/browse/PIMOB-1906)

### CODE REVIEW

**Comment by alex-ioja-yang-cko** [comment]
Come on GitHub, see the commits pushed while you were having issues 🙏

File: Tests/UI/BillingForm/ViewModel/CardCreationModelTests.swift [hunk 1]
```diff
@@ -85,4 +85,44 @@ final class CardCreationModelTests: XCTestCase {
         XCTAssertEqual(card?.billingAddress, model.billingAddress)
         XCTAssertEqual(card?.phone, model.phone)
     }
+    
+    func testCVVOptionalTrue() {
```
Comment (deepesh-vasthimal-cko) [inline Tests/UI/BillingForm/ViewModel/CardCreationModelTests.swift:89]: Can you rename the function to be more meaningful having Suffix True is minimal I feel.

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift [hunk 2]
```diff
@@ -16,17 +16,34 @@ struct CardCreationModel {
...
                 billingAddress: billingAddress,
                 phone: phone)
   }
+
+  private func submissionCVV() -> String? {
+    // If CVV is provided, we will submit it
+    if !cvv.isEmpty { return cvv }
+
+    // If CVV was not provided and the field is optional, return nil
+    if isCVVOptional { return nil}
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:43]:
```suggestion
    if isCVVOptional { return nil }
```

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift [hunk 3]
```diff
@@ -16,17 +16,34 @@ struct CardCreationModel {
...
+
+    // If CVV was not optional, we return the empty value we
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:45]:
```suggestion
    // If CVV was not optional, we return the empty value.
```

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift [hunk 4]
```diff
@@ -16,17 +16,34 @@ struct CardCreationModel {
...
                 expiryDate: expiryDate,
                 name: name,
-                cvv: cvv,
+                cvv: submissionCVV(),
                 billingAddress: billingAddress,
                 phone: phone)
   }
+
+  private func submissionCVV() -> String? {
+    // If CVV is provided, we will submit it
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:39]: Are we submitting the CVV? Can you rephrase this sentence?

File: Source/UI/PaymentForm/ViewModel/DefaultPaymentViewModel.swift [hunk 5]
```diff
@@ -43,6 +43,8 @@ class DefaultPaymentViewModel: PaymentViewModel {
         self.paymentFormStyle = paymentFormStyle
         self.billingFormStyle = billingFormStyle
         self.logger = logger
+        let isCVVOptional = paymentFormStyle?.securityCode == nil
```
Comment (deepesh-vasthimal-cko) [inline Source/UI/PaymentForm/ViewModel/DefaultPaymentViewModel.swift:46]: Does this mean if a user does not enter a CVV then we consider that to be CVV optional by default?

Comment (alex-ioja-yang-cko) [reply Source/UI/PaymentForm/ViewModel/DefaultPaymentViewModel.swift:46]: This looks at the formStyle, not the user input. The form is provided by the SDK consumer (application), not the user.

What this line of code means is that if the app developer chooses to hide CVV field, our ParsingModel for the inputs (CardCreationModel) will treat its presence as optional. And since there is no UI to enter it, it will work without it.
This allows us if we want to follow Web format of (Show & optional), to furthermore add a property on the actual style that will allow us to show & have optional.

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift [hunk 6]
```diff
@@ -16,17 +16,34 @@ struct CardCreationModel {
...
     return Card(number: number,
                 expiryDate: expiryDate,
                 name: name,
-                cvv: cvv,
+                cvv: submissionCVV(),
                 billingAddress: billingAddress,
                 phone: phone)
   }
+
+  private func submissionCVV() -> String? {
```
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:38]: I don't really know how to rephrase it, as it gets the CVV available to this parser object and applies business logic to get a final CVV that will be submitted via the Card object to network. 

Hence, I am a little stuck on `submissionCVV` as its the CVV variant that will get submitted 🤔

**Review by deepesh-vasthimal-cko — APPROVED** [review]

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift [hunk 7]
```diff
@@ -16,17 +16,34 @@ struct CardCreationModel {
...
+    // If CVV is provided, we will return it
+    if !cvv.isEmpty { return cvv }
+
+    // If CVV was not provided and the field is optional, return nil
+    if isCVVOptional { return nil }
+
+    // If CVV was not optional, we return the CVV as it is (empty string)
+    return cvv
```
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:46]: Your suggestion would ignore any value of CVV that was entered. The current code as described purposely uses any value of CVV that may have been introduced.

Whilst in this version entering CVV is not possible, the behaviour does not refer to a hidden/show state, but rather an optional/required state. Currently only hidden triggers optional, but if we were to match web, the optional would be extended to also handle "client requested"

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift (hunk 6 above)
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:38]: I am afraid this is not validating the CVV, it decides whether what value to submit.

File: Tests/UI/BillingForm/ViewModel/PaymentViewModelTests.swift [hunk 8]
```diff
@@ -652,7 +652,7 @@ final class PaymentViewModelTests: XCTestCase {
         
         model.payButtonIsPressed()
         
-        let expectedPaymentCard = Card(number: cardNumber, expiryDate: expiryDate, name: "", cvv: "", billingAddress: nil, phone: nil)
+        let expectedPaymentCard = Card(number: cardNumber, expiryDate: expiryDate, name: "", cvv: nil, billingAddress: nil, phone: nil)
```
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Tests/UI/BillingForm/ViewModel/PaymentViewModelTests.swift:655]: I don't understand what you mean? Name accepts a String, but its value is irrelevant to the unit test purpose?

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift [hunk 9]
```diff
@@ -16,17 +16,34 @@ struct CardCreationModel {
   var phone: Phone?
   var scheme: Card.Scheme?
 
+  let isCVVOptional: Bool
```
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:19]: In this particular case I do not agree. The empty line in this case is helping set a clear differentiator between different behaviours of the properties of the object. In more complex objects such a differentiator may become noise, but in a simple example like this it allows the reader to quickly notice the change in behaviour. Full context of object (beyond what code comparison shows) is:

```Swift
struct CardCreationModel {
  var number: String = ""
  var expiryDate: ExpiryDate?
  var name: String = ""
  var cvv: String = ""
  var billingAddress: Address?
  var phone: Phone?
  var scheme: Card.Scheme?

   let isCVVOptional: Bool
}
```

Any reviewer trying to understand what this object does could quickly discover all the mutable properties that they can synchronise, whilst also seeing the immutable nature of `isCVVOptional` which cannot be modified after creation.

File: Source/UI/PaymentForm/ViewModel/CardCreationModel.swift (hunk 7 above)
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Source/UI/PaymentForm/ViewModel/CardCreationModel.swift:46]:
```Swift
let isCVVOptional = true
let cvv = "123"

guard !cvv.isEmpty || !isCVVOptional else {
    return nil
}
return cvv
```
Using your code suggestion and trying simple values we have ☝️. That would return nil, not "123", which results in ignoring the value.

File: Tests/UI/BillingForm/ViewModel/PaymentViewModelTests.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (alex-ioja-yang-cko) [reply Tests/UI/BillingForm/ViewModel/PaymentViewModelTests.swift:655]: In that case I believe I have already responded with
> Name accepts a String, but its value is irrelevant to the unit test purpose

Do note this is a unit test beyond the scope of this PR. The fact that the CVV had to be updated to `nil` shows the behaviour intended is achieved (if secureCode is not displayed, the model will create a Card object with `nil` as CVV)
